src/server/ua_server_ns0_pushmanagement.c

```diff
@@ -0,0 +1,1126 @@
+    fileContext->currentPos = 0;
+    retval = createFileHandleId(fileInfo, &fileContext->fileHandle);
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_ByteString_delete(fileContext->file);
+        UA_free(fileContext);
+        if(fileInfo->openCount == 0) {
+            UA_free(fileInfo);
+            certGroup->applicationContext = NULL;
+        }
+        return retval;
+    }
+
+    UA_FileContext *fileContextTmp = NULL;
+    LIST_FOREACH(fileContextTmp, &fileInfo->fileContext, listEntry) {
+        if(fileContextTmp->openFileMode & (UA_OPENFILEMODE_WRITE | UA_OPENFILEMODE_ERASEEXISTING))
+            retval = UA_STATUSCODE_BADNOTREADABLE;
+    }
+
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_ByteString_delete(fileContext->file);
+        UA_free(fileContext);
+        if(fileInfo->openCount == 0) {
+            UA_free(fileInfo);
+            certGroup->applicationContext = NULL;
+        }
+        return retval;
+    }
+
+    fileInfo->openCount += 1;
+    UA_Variant_setScalarCopy(output, &fileContext->fileHandle, &UA_TYPES[UA_TYPES_UINT32]);
+
+    LIST_INSERT_HEAD(&fileInfo->fileContext, fileContext, listEntry);
+
+    return UA_STATUSCODE_GOOD;
+}
+
+static UA_StatusCode
+readTrustList(UA_Server *server,
+              const UA_NodeId *sessionId, void *sessionHandle,
+              const UA_NodeId *methodId, void *methodContext,
+              const UA_NodeId *objectId, void *objectContext,
+              size_t inputSize, const UA_Variant *input,
+              size_t outputSize, UA_Variant *output) {
+
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+
+    /*check for input types*/
+    if(!UA_Variant_hasScalarType(&input[0], &UA_TYPES[UA_TYPES_UINT32]) || /*FileHandle*/
+       !UA_Variant_hasScalarType(&input[1], &UA_TYPES[UA_TYPES_INT32])) /*Length*/
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+
+    UA_UInt32 fileHandle = *(UA_UInt32*)input[0].data;
+    UA_Int32 length = *(UA_Int32*)input[1].data;
+
+    UA_CertificateGroup *certGroup = getCertGroup(server, objectId);
+    if(certGroup == NULL) {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+
+    if(certGroup->applicationContext == NULL)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    UA_FileInfo *fileInfo = (UA_FileInfo*)certGroup->applicationContext;
+    UA_FileContext *fileContext = getFileContext(fileInfo, sessionId, fileHandle);
+
+    if(fileContext == NULL)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    if(fileContext->openFileMode & (UA_OPENFILEMODE_WRITE | UA_OPENFILEMODE_ERASEEXISTING))
+        return UA_STATUSCODE_BADINVALIDSTATE;
+
+    /* check boundaries */
+    if((size_t)length >= fileContext->file->length) {
+        length = fileContext->file->length;
+    }
+    if((size_t)length >= (fileContext->file->length - fileContext->currentPos)) {
+        length = (fileContext->file->length - fileContext->currentPos);
+    }
+
+    UA_ByteString *readBuffer = UA_ByteString_new();
```

Comment:
This can be done easier once the length is correct.
```c
UA_ByteString localCopy = fileContext->file;
localCopy.length = length;
localCopy.data += fileContext->currentPos;
UA_ByteString_copy(&localCopy, readBuffer);
fileContext->currentPos += length;
```

And of course check the result of the _copy as well.